chrononaut/versioned.py

```diff
@@ -70,113 +77,125 @@ def versions(self, before=None, after=None, return_query=False):
-            raise ChrononautException("Cannot diff from a non-history model.")
+        if not isinstance(from_timestamp, datetime):
+            raise ChrononautException("The diff method takes datetime as its argument.")
 
-        if to_model is not self and from_model.changed > to_model.changed:
+        if to_timestamp < from_timestamp:
             raise ChrononautException(
-                "Diffs must be chronological. Your from_model " "post-dates your to."
+                "Diffs must be chronological. Your from_model post-dates your to."
             )
 
-        # TODO: Refactor this and `create_version` so some of the object mapper
-        #       iteration is not duplicated twice
+        from_dict = self.version_at(from_timestamp, return_snapshot=True)
+        to_dict = self.version_at(to_timestamp, return_snapshot=True)
+
+        hidden_cols = set(getattr(self, "__chrononaut_hidden__", []))
+        all_keys = set(from_dict.keys())
+        all_keys.update(to_dict.keys())
+        all_keys = all_keys.difference(hidden_cols)
+
         diff = {}
-        obj_mapper = object_mapper(from_model)
-        for om in obj_mapper.iterate_to_root():
-            for obj_col in om.local_table.c:
-                if "version_meta" in obj_col.info or obj_col.key in untracked_cols:
-                    continue
-                try:
-                    prop = obj_mapper.get_property_by_column(obj_col)
-                except UnmappedColumnError:
-                    continue
-
-                # First check the history model's columns
-                from_val = getattr(from_model, prop.key)
-                to_val = getattr(to_model, prop.key)
-                if from_val != to_val:
-                    diff[prop.key] = (from_val, to_val)
+        for k in all_keys:
+            if k in from_dict and k not in to_dict:
+                diff[k] = (from_dict[k], None)
+            elif k not in from_dict and k in to_dict:
+                diff[k] = (None, to_dict[k])
+            else:
+                # it's in both
+                if from_dict[k] != to_dict[k]:
+                    diff[k] = (from_dict[k], to_dict[k])
 
         # If `include_hidden` we need to enumerate through every
-        # model *since* the from_model and see if `change_info` includes
-        # hidden columns. We only need to do this for non-history instances.
-        if include_hidden and isinstance(to_model, self.__class__):
-            from_versions = self.versions(after=from_model.changed)
-            for from_version in from_versions:
-                if "hidden_cols_changed" in from_version.change_info:
-                    for hidden_col in from_version.change_info["hidden_cols_changed"]:
-                        diff[hidden_col] = (None, getattr(to_model, hidden_col))
+        # model *since* the from_timestamp *until* the to_timestamp
+        # and see if `extra_info` includes hidden columns.
```

Comment:
I don't understand what `include_hidden` means, just checking for changes in columns now hidden but that weren't before? Do we actually use that anywhere @boydgreenfield ?

Reply:
This param is for including hidden column changes in the diff. We record that hidden columns changed between versions, but don't record explicit values. By default we only list explicit value changes in the diff, but this param allows for adding information about hidden columns, e.g. user changed their password between versions

Reply:
@pkonieczny how does a password change get stored under the new scheme?

Reply:
If the column is marked as hidden, only the column name is listed among changed columns in extra info